### PR TITLE
Feature/#196 list citizens

### LINKED ISSUE #196: List citizens
We should have a way to list the citizen.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #196 

# Changes proposed in this pull request:
- Allows to list citizens

Review please @Minecolonies/maintainers

### CODE REVIEW

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    @NotNull
+    @Override
+    public String getCommandUsage(@NotNull final ICommandSender sender)
+    {
+        return super.getCommandUsage(sender) + "";
+    }
+
+    @Override
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MAJOR This method has 101 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
MAJOR The Cyclomatic Complexity of this method "execute" is 16 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/MineColonies.java [hunk 2]
```diff
@@ -164,6 +164,7 @@ public void serverLoad(FMLServerStartingEvent event)
     {
         // register server commands
         event.registerServerCommand(new MinecoloniesCommand());
+
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/MineColonies.java]: why?

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 1 above)
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: Definitely fix that, I will work on a better paging framework next...

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                    //ignore and keep page 1.
+                }
+            }
+        }
+
+        final Colony colony = ColonyManager.getColony(colonyId);
+
+        if(colony == null)
+        {
+            return;
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: feedback for the user?

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    @NotNull
+    @Override
+    public List<String> getTabCompletionOptions(
+                                                 @NotNull final MinecraftServer server,
+                                                 @NotNull final ICommandSender sender,
+                                                 @NotNull final String[] args,
+                                                 @Nullable final BlockPos pos)
+    {
+        return new ArrayList<>();
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: We could suggest the players colony (or colonies) here?

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        int page = 1;
+        int colonyId = 1;
+
+        final IColony tempColony = ColonyManager.getIColonyByOwner(sender.getEntityWorld(), sender.getCommandSenderEntity().getUniqueID());
+        if(tempColony != null)
+        {
+            final Colony colony = ColonyManager.getColony(sender.getEntityWorld(), tempColony.getCenter());
+            if(colony != null)
+            {
+                colonyId = colony.getID();
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: this behavior is convenient but I think the structure gets confusing...

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                {
+                    colonyId = Integer.parseInt(args[0]);
+                    page = Integer.parseInt(args[1]);
+                }
+                catch (NumberFormatException e)
+                {
+                    //ignore and keep page 1.
+                }
+            }
+            else
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: if we keep this style I have to do something about argumentparsing... thats looks horrible 😆 Will do that after the other two

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 5 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: Yeah I'll put it in an extra method.

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 6 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: We should do a real argument parser.

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 4 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: I don't know how that method is supposed to work.

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    @Override
+    public String getCommandUsage(@NotNull final ICommandSender sender)
+    {
+        return super.getCommandUsage(sender) + "";
+    }
+
+    @Override
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
+    {
+        int colonyId = getIthArgument(args, 0, getColonyId(sender));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MINOR Local variable 'colonyId' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * Draws the page switcher at the bottom
+     * @param sender the sender.
+     * @param colonyId the colonyid.
+     * @param page the page number.
+     * @param count number of citizens.
+     * @param halfPage the halfPage.
+     */
+    private static void drawPageSwitcher(@NotNull final ICommandSender sender, int colonyId, int page, int count, int halfPage)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MAJOR Remove this unused method parameter "colonyId". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1172)

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    /**
+     * Get the ith argument (An Integer).
+     * @param i the argument from the list you want.
+     * @param args the list of arguments.
+     * @param def the default value.
+     * @return the argument.
+     */
+    private int getIthArgument(String[] args, int i, int def)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MINOR Make "getIthArgument" a "static" method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            if(colony != null)
+            {
+                return colony.getID();
+            }
+        }
+
+        return 1;
+    }
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+
+    /**
+     * Returns the colony of the owner or if not available colony 1.
+     * First tries to get the IColony and then the Colony from the ColonyManager.
+     * @param sender the sender of the command.
+     * @return the colonyId.
+     */
+    private int getColonyId( @NotNull final ICommandSender sender)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: MINOR Make "getColonyId" a "static" method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2325)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 5 issues
* MAJOR 1 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by Raycoms** [comment]
@Minecolonies/maintainers 

Can we look over this, we need these commands on the server.

File: src/main/java/com/minecolonies/commands/CitizensCommand.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+ * Minecolonies root command.
+ * <p>
+ * Manages all sub commands.
+ */
+public class CitizensCommand extends AbstractSplitCommand
+{
+
+    private final ImmutableMap<String, ISubCommand> subCommands =
+      new ImmutableMap.Builder<String, ISubCommand>()
+        .put("list", new ListCitizens("mc", "citizens", "list"))
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/CitizensCommand.java]: We should use DESC here too

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 7 above)
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: please do

File: src/main/java/com/minecolonies/commands/ListCitizens.java [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+    public String getCommandUsage(@NotNull final ICommandSender sender)
+    {
+        return super.getCommandUsage(sender) + "";
+    }
+
+    @Override
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
+    {
+        int colonyId = getIthArgument(args, 0, getColonyId(sender));
+        final Colony colony = ColonyManager.getColony(colonyId);
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: still, why do we need the actual colony?

File: src/main/java/com/minecolonies/commands/ListCitizens.java (hunk 4 above)
Comment (Kostronor) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: It gets (like execute) a half done command (when the user hits tab) and it has to return a list of strings completing the last argument. Each String is an auto complete option available to the player

Comment (Raycoms) [inline src/main/java/com/minecolonies/commands/ListCitizens.java]: I think suggesting numbers wouldn't really help the player, may confuse him more.

**Pull request closed without merging**